File: packages/reown_yttrium/ios/Classes/ChainAbstraction.swift
import Flutter
import UIKit
import YttriumWrapper

class ChainAbstraction {
    
    struct CustomError: Error {
        let message: String
    }
    
    private static var chainAbstractionClient: ChainAbstractionClient?
    private static var pendingPrepareDetailed: [String: UiFields] = [:]
    
    static func initialize(_ params: Any, result: @escaping FlutterResult) {
        guard let dict = params as? [String: Any],
              let projectId = dict["projectId"] as? String,
              let pulseMetadataDict = dict["pulseMetadata"] as? [String: Any],
              let url = pulseMetadataDict["url"] as? String,
              let bundleId = pulseMetadataDict["bundleId"] as? String,
              let sdkVersion = pulseMetadataDict["sdkVersion"] as? String,
              let sdkPlatform = pulseMetadataDict["sdkPlatform"] as? String else {
            result(FlutterError(code: "initialize", message: "Invalid parameters", details: params))
            return
        }
        
        let bundleID = Bundle.main.bundleIdentifier ?? bundleId
        let pulseMetadata = PulseMetadata(url: url, bundleId: bundleID, sdkVersion: sdkVersion, sdkPlatform: sdkPlatform)
        
        chainAbstractionClient = ChainAbstractionClient(projectId: projectId, pulseMetadata: pulseMetadata)
        
        result(true)
    }
    
    static func erc20TokenBalance(_ params: Any, result: @escaping FlutterResult) {
        print("erc20TokenBalance called with ", params)
        
        guard let dict = params as? [String: Any],
              let chainId = dict["chainId"] as? String,
              let token = dict["token"] as? FfiAddress,  // aka String
              let owner = dict["owner"] as? FfiAddress,
              let client = chainAbstractionClient else {
            result(FlutterError(code: "erc20TokenBalance", message: "Invalid parameters", details: params))
            return
        }
        
        Task {
            do {
                let balanceResponse = try await client.erc20TokenBalance(chainId: chainId, token: token, owner: owner)
                
                if let balance = balanceResponse as String? {
                    print("balanceResponse", balance)
                    result(balance)
                } else {
                    result(FlutterError(code: "erc20TokenBalance", message: "Response error: balance is nil", details: nil))
                }
            } catch {
                result(FlutterError(code: "erc20TokenBalance", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    static func estimateFees(_ params: Any, result: @escaping FlutterResult) {
        print("estimateFees called with ", params)
        
        guard let dict = params as? [String: Any],
              let chainId = dict["chainId"] as? String,
              let client = chainAbstractionClient else {
            result(FlutterError(code: "estimateFees", message: "Invalid parameters", details: params))
            return
        }
        
        Task {
            do {
                let feesResponse = try await client.estimateFees(chainId: chainId)
                
                if let fees = feesResponse as Eip1559Estimation? {
                    print("feesResponse", fees)
                    result(fees.toJson())
                } else {
                    result(FlutterError(code: "estimateFees", message: "Response error: fees is nil", details: nil))
                }
            } catch {
                result(FlutterError(code: "estimateFees", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    static func prepareDetailed(_ params: Any, result: @escaping FlutterResult) {
        print("prepareDetailed called with ", params)
        
        guard let dict = params as? [String: Any],
              let chainId = dict["chainId"] as? String,
              let from = dict["from"] as? String,
              let accounts = dict["accounts"] as? [String],
              let useLifi = dict["useLifi"] as? Bool,
              let call = dict["call"] as? [String: Any],
              let to = call["to"] as? String,  // Address (String)
              let value = call["value"] as? String,  // U256 (String)
              let input = call["input"] as? String,  // Bytes (String)
              let localCurrency = dict["localCurrency"] as? String,
              let client = chainAbstractionClient else {
            result(FlutterError(code: "prepareDetailed", message: "Invalid parameters", details: nil))
            return
        }
        
        Task {
            do {
                let currency = Currency.fromString(value: localCurrency)
                let call = Call(to: to, value: value, input: input)
                let prepareResponse = try await client.prepareDetailed(chainId: chainId,
                                                                                       from: from,
                                                                                       call: call,
                                                                                       accounts: accounts,
                                                                                       localCurrency: currency,
                                                                                       useLifi: useLifi)
                
                switch prepareResponse {
                case .success(let successData):
                    switch successData {
                    case .available(let uiFields):
                        pendingPrepareDetailed[uiFields.routeResponse.orchestrationId] = uiFields
                        result(["available": uiFields.toJson()])
                    case .notRequired(let prepareResponseNotRequired):
                        result(["notRequired": prepareResponseNotRequired.toJson()])
                    }
                case .error(let errorData):
                    result(errorData.toJson())
                }
            } catch {
                result(FlutterError(code: "prepareDetailed", message: error.localizedDescription, details: nil))
            }
        }
    }
    
    static func execute(_ params: Any, result: @escaping FlutterResult) {
        print("execute called with ", params)
        
        guard let dict = params as? [String: Any],
              let orchestrationId = dict["orchestrationId"] as? String,
              let rawSigs = dict["routeTxnSigs"] as? [Any],
              let initialTxnSig = dict["initialTxnSig"] as? String,
              let client = chainAbstractionClient else {
            result(FlutterError(code: "execute", message: "Invalid parameters", details: params))
            return
        }
        
        Task {
            do {
                guard let uiFields = pendingPrepareDetailed[orchestrationId] else {
                    throw CustomError(message: "prepareDetailed result not found, try again")
                }
                
                // Try converting to [RouteSig]
                let routeTxnSigs: [RouteSig] = rawSigs.compactMap { $0 as? RouteSig }
                if routeTxnSigs.count != rawSigs.count {
                    result(FlutterError(code: "execute", message: "Invalid item(s) in routeTxnSigs", details: params))
                    return
                }

                
                
                let executeResponse = try await client.execute(uiFields: uiFields,
                                                               routeTxnSigs: routeTxnSigs,
                                                               initialTxnSig: initialTxnSig)
                
                pendingPrepareDetailed.removeValue(forKey: orchestrationId)
                
                let jsonResponse = executeResponse.toJson()
                print("execute success", jsonResponse)
                result(jsonResponse)
            } catch {
                result(FlutterError(code: "execute", message: error.localizedDescription, details: nil))
            }
        }
    }
}

File: packages/reown_yttrium/android/src/main/kotlin/com/reown/reown_yttrium/ChainAbstraction.kt
package com.reown.reown_yttrium

import android.content.Context
import com.yttrium.YttriumKt
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import uniffi.uniffi_yttrium.ChainAbstractionClient
import uniffi.uniffi_yttrium.Eip1559Estimation
import uniffi.yttrium.Call
import uniffi.yttrium.Currency
import uniffi.yttrium.PrepareDetailedResponse
import uniffi.yttrium.PrepareDetailedResponseSuccess
import uniffi.yttrium.PulseMetadata
import uniffi.yttrium.UiFields
import kotlin.collections.set

class ChainAbstraction {
    companion object {
        private lateinit var chainAbstractionClient: ChainAbstractionClient
        private var pendingPrepareDetailed: MutableMap<String, UiFields> = mutableMapOf()

        fun initialize(applicationContext: Context, params: Any?, result: MethodChannel.Result) {
//            YttriumKt.initializeTls(applicationContext)

            val dict = params as? Map<*, *> ?: return result.error("initialize", "Invalid parameters: not a map", null)

            val projectId = dict["projectId"] as? String ?: return errorMissing("projectId", params, result)
            val pulseMetadataDict = dict["pulseMetadata"] as? Map<*, *> ?: return errorMissing("pulseMetadata", params, result)

            val url = pulseMetadataDict["url"] as? String ?: return errorMissing("url", params, result)
            val packageName = applicationContext.packageName ?: pulseMetadataDict["packageName"] as? String ?: return errorMissing("packageName", params, result)
            val sdkVersion = pulseMetadataDict["sdkVersion"] as? String ?: return errorMissing("sdkVersion", params, result)
            val sdkPlatform = pulseMetadataDict["sdkPlatform"] as? String ?: return errorMissing("sdkPlatform", params, result)

            val pulseMetadata = PulseMetadata(url, packageName, sdkVersion, sdkPlatform)

            chainAbstractionClient = ChainAbstractionClient(projectId = projectId, pulseMetadata = pulseMetadata)

            result.success(true)
        }

        fun erc20TokenBalance(params: Any?, result: MethodChannel.Result) {
            println("erc20TokenBalance called with $params")

            val dict = params as? Map<*, *> ?: return result.error("erc20TokenBalance", "Invalid parameters: not a map", null)

            val tokenAddress = dict["token"] as? String ?: return errorMissing("token", params, result)
            val ownerAddress = dict["owner"] as? String ?: return errorMissing("owner", params, result)
            val chainId = dict["chainId"] as? String ?: return errorMissing("chainId", params, result)

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val balanceResponse = chainAbstractionClient.erc20TokenBalance(chainId, tokenAddress, ownerAddress)
                    println("erc20TokenBalance response $balanceResponse")
                    result.success(balanceResponse)
                } catch (e: Exception) {
                    result.error("erc20TokenBalance", "Yttrium erc20TokenBalance Error: ${e.message}", null)
                }
            }
        }

        fun estimateFees(params: Any?, result: MethodChannel.Result) {
            println("estimateFees called with $params")

            (params as? Map<*, *>)?.let { dict ->
                val chainId = dict["chainId"] as? String

                if (chainId != null) {
                    CoroutineScope(Dispatchers.IO).launch {
                        try {
                            val feesResponse: Eip1559Estimation = chainAbstractionClient.estimateFees(chainId)
                            println("estimateFees response $feesResponse")
                            result.success(feesResponse.toMap())
                        } catch (e: Exception) {
                            result.error("estimateFees", "Yttrium estimateFees Error: ${e.message}", null)
                        }
                    }
                    return
                }
            }

            result.error("estimateFees", "Invalid parameters $params", null)
        }

        fun prepareDetailed(params: Any?, result: MethodChannel.Result) {
            println("prepareDetailed called with $params")

            val dict = params as? Map<*, *> ?: return result.error("prepareDetailed", "Invalid parameters: not a map", null)

            val chainId = dict["chainId"] as? String ?: return errorMissing("chainId", params, result)
            val from = dict["from"] as? String ?: return errorMissing("from", params, result)
            val accounts = (dict["accounts"] as? List<*>)?.filterIsInstance<String>() ?: return errorMissing("accounts", params, result)
            val call = dict["call"] as? Map<*, *> ?: return errorMissing("call", params, result)
            val to = call["to"] as? String ?: return errorMissing("call.to", params, result)
            val value = call["value"] as? String ?: return errorMissing("call.value", params, result)
            val input = call["input"] as? String ?: return errorMissing("call.input", params, result)
            val localCurrency = dict["localCurrency"] as? String ?: return errorMissing("localCurrency", params, result)
            val useLifi = dict["useLifi"] as? Boolean ?: return errorMissing("useLifi", params, result)

            CoroutineScope(Dispatchers.IO).launch {
                val response = chainAbstractionClient.prepareDetailed(
                    chainId = chainId,
                    from = from,
                    accounts = accounts,
                    call = Call(
                        to = to,
                        value = value,
                        input = input
                    ),
                    localCurrency = Currency.valueOf(localCurrency.uppercase()),
                    useLifi = useLifi,
                )

                when (response) {
                    is PrepareDetailedResponse.Success -> {
                        when (val success = response.v1) {
                            is PrepareDetailedResponseSuccess.Available -> {
                                val uiFields = success.v1
                                pendingPrepareDetailed[uiFields.routeResponse.orchestrationId] = uiFields
                                println("prepareDetailed response $uiFields")
                                result.success(mapOf("available" to uiFields.toMap()))
                            }
                            is PrepareDetailedResponseSuccess.NotRequired -> {
                                val notRequired = success.v1
                                println("prepareDetailed notRequired $notRequired")
                                result.success(mapOf("notRequired" to notRequired.toMap()))
                            }
                        }
                    }
                    is PrepareDetailedResponse.Error -> {
                        println("prepareDetailed: error -> ${response.v1}")
                        result.success(response.v1.toMap())
                    }
                }
            }
        }

        fun execute(params: Any?, result: MethodChannel.Result) {
            println("execute called with $params")

            val dict = params as? Map<*, *> ?: return result.error("execute", "Invalid parameters: not a map", null)

            val orchestrationId = dict["orchestrationId"] as? String ?: return errorMissing("orchestrationId", params, result)
            val routeTxnSigs = (dict["routeTxnSigs"] as? List<*>)?.filterIsInstance<uniffi.yttrium.RouteSig>()
                ?: return errorMissing("routeTxnSigs", params, result)
            val initialTxnSig = dict["initialTxnSig"] as? String ?: return errorMissing("initialTxnSig", params, result)

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val prepareDetailedResult = pendingPrepareDetailed[orchestrationId]
                    val executionResult = prepareDetailedResult?.let {
                        chainAbstractionClient.execute(
                            uiFields = it,
                            routeTxnSigs = routeTxnSigs,
                            initialTxnSig = initialTxnSig
                        )
                    }

                    println("execute response $executionResult")
                    pendingPrepareDetailed.remove(orchestrationId)
                    result.success(executionResult?.toMap())
                } catch (e: Exception) {
                    println("execute error: $e")
                    result.error("execute", "Yttrium execute Error: ${e.message}", null)
                }
            }
        }

    }
}